check_diff/src/lib.rs

```diff
@@ -56,3 +153,143 @@ pub fn change_directory_to_path(dest: &Path) -> io::Result<()> {
+    let Ok(command) = Command::new("cargo").args(["--version"]).output() else {
+        return Err(CheckDiffError::FailedCargoVersion(
+            "Failed to obtain cargo version".to_string(),
+        ));
+    };
+
+    let Ok(cargo_version) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting cargo version to string".to_string(),
+        ));
+    };
+
+    return Ok(cargo_version);
+}
+
+pub fn get_binary_version(
+    binary: &PathBuf,
+    ld_lib_path: &String,
+) -> Result<String, CheckDiffError> {
+    let binary_str = pathbuf_to_str(binary)?;
+    let Ok(command) = Command::new(binary_str)
+        .env("LD_LIB_PATH", ld_lib_path)
+        .args(["--version"])
+        .output()
+    else {
+        return Err(CheckDiffError::FailedVersioning(format!(
+            "Failed to get version for {:?}",
+            binary
+        )));
+    };
+
+    let Ok(binary_version) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting binary version to string".to_string(),
+        ));
+    };
+    return Ok(binary_version);
+}
+
+/// Obtains the ld_lib path and then builds rustfmt from source
+/// If that operation succeeds, the source is then copied to the output path specified
+pub fn build_rustfmt_from_src(output_path: &PathBuf) -> Result<RustfmtRunner, CheckDiffError> {
+    //Because we're building standalone binaries we need to set `LD_LIBRARY_PATH` so each
+    // binary can find it's runtime dependencies.
+    // See https://github.com/rust-lang/rustfmt/issues/5675
+    // This will prepend the `LD_LIBRARY_PATH` for the master rustfmt binary
+    let ld_lib_path = get_ld_lib_path()?;
+
+    info!("Building rustfmt from source");
+    let Ok(_) = Command::new("cargo")
+        .env("LD_LIB_PATH", &ld_lib_path.as_str())
+        .args(["build", "-q", "--release", "--bin", "rustfmt"])
+        .output()
+    else {
+        return Err(CheckDiffError::FailedSourceBuild(
+            "Error building rustfmt from source".to_string(),
+        ));
+    };
+
+    std::fs::copy("target/release/rustfmt", &output_path)?;
+
+    return Ok(RustfmtRunner {
+        ld_library_path: ld_lib_path,
+        binary_path: output_path.to_path_buf(),
+    });
+}
+
+fn pathbuf_to_str(pathbuf: &PathBuf) -> Result<&str, CheckDiffError> {
+    let Some(result) = pathbuf.to_str() else {
+        return Err(CheckDiffError::FailedPathBuf(format!(
+            "Unable to convert {:?} to str",
+            pathbuf
+        )));
+    };
+    return Ok(result);
+}
+
+// Compiles and produces two rustfmt binaries.
+// One for the current master, and another for the feature branch
+// Parameters:
+// dest: Directory where rustfmt will be cloned
+pub fn compile_rustfmt(
+    dest: &Path,
+    remote_repo_url: String,
+    feature_branch: String,
+    commit_hash: Option<String>,
+) -> Result<[RustfmtRunner; 2], CheckDiffError> {
```

Comment:
I was deciding between returning a fixed sized array and a vector of `RustfmtRunner`s.